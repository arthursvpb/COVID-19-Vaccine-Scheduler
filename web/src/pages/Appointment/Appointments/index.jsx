/* eslint-disable no-underscore-dangle */
import React, { useEffect, useState } from 'react';

import DatePicker from 'react-datepicker';

import EventIcon from '@material-ui/icons/Event';

// import { makeStyles } from '@material-ui/core/styles';
import {
  Paper,
  TextField,
  InputAdornment,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Grid,
  Typography,
} from '@material-ui/core';

import { parse, differenceInCalendarYears, format } from 'date-fns';

import Image from 'material-ui-image';
import undrawNoData from '../../../assets/undraw_no_data.svg';

import Page from '../../../components/Page';
import Button from '../../../components/Button';

import api from '../../../services/api';

import useStyle from './style';

export default function index() {
  const [appointments, setAppointments] = useState([]);
  const [filteredAppointments, setFilteredAppointments] = useState([]);
  const [dateFilter, setDateFilter] = useState('');

  const [showDialog, setShowDialog] = useState(false);
  const [conclusionInput, setConclusionInput] = useState('');

  const [editAppointmentId, setEditAppointmentId] = useState();

  useEffect(() => {
    api
      .get('/appointments')
      .then(response => {
        setAppointments(response.data);
      })
      .catch(error => {
        alert(
          error.response?.data.message ||
            `ðŸ˜“ Something went wrong! Is server down?`,
        );
      });
  }, [filteredAppointments]);

  const handleDateFilterChange = date => {
    const dateString = format(date, 'dd-MM-yyyy');

    setFilteredAppointments(
      appointments.filter(
        ({ vaccinationDate }) => vaccinationDate === dateString,
      ),
    );

    setDateFilter(date);
  };

  const handleDialogOpen = _id => {
    setEditAppointmentId(_id);
    setShowDialog(true);
  };

  const handleDialogClose = () => {
    setShowDialog(false);
  };

  const handleConclusionSubmit = async event => {
    event.preventDefault();

    const updatedAppointments = filteredAppointments.map(appointment => {
      if (appointment._id === editAppointmentId) {
        return {
          ...appointment,
          isConcluded: true,
          conclusion: conclusionInput,
        };
      }

      return appointment;
    });

    try {
      await api.put(`/appointments/${editAppointmentId}`, {
        isConcluded: true,
        conclusion: conclusionInput,
      });

      setFilteredAppointments(updatedAppointments);
      handleDialogClose();
      alert(`âœ… Success!`);
    } catch (error) {
      alert(error.response?.data.message || `ðŸ˜“ Something went wrong!`);
    }
  };

  const classes = useStyle();

  return (
    <Page
      style={{
        flexDirection: 'column',
        justifyContent: 'flex-start',
      }}
    >
      <Grid container className={classes.grid}>
        <Grid item xs={12} className={classes.title}>
          <DatePicker
            customInput={
              <TextField
                inputProps={{
                  style: { textAlign: 'center', fontSize: '20px' },
                }}
                // eslint-disable-next-line react/jsx-no-duplicate-props
                InputProps={{
                  startAdornment: (
                    <InputAdornment position="start">
                      <EventIcon />
                    </InputAdornment>
                  ),
                  endAdornment: <InputAdornment position="end" />,
                }}
                label="Data da vacinaÃ§Ã£o"
                variant="outlined"
              />
            }
            name="dateFilter"
            id="dateFilter"
            selected={dateFilter}
            dateFormat="dd-MM-yyyy"
            onChange={handleDateFilterChange}
          />
        </Grid>
        <Grid item xs={12}>
          {filteredAppointments.length > 0 ? (
            <TableContainer component={Paper}>
              <Table>
                <TableHead>
                  <TableRow>
                    <TableCell align="center">Nome</TableCell>
                    <TableCell align="center">Idade</TableCell>
                    <TableCell align="center">HorÃ¡rio</TableCell>
                    <TableCell align="center">Status</TableCell>
                    <TableCell align="center">ConclusÃ£o</TableCell>
                    <TableCell align="center">AÃ§Ãµes</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {filteredAppointments.map(
                    ({
                      _id,
                      name,
                      birthday,
                      vaccinationTime,
                      conclusion,
                      isConcluded,
                    }) => (
                      <TableRow key={_id}>
                        <TableCell align="center">{name}</TableCell>
                        <TableCell align="center">
                          {differenceInCalendarYears(
                            new Date(),
                            parse(birthday, 'dd-MM-yyyy', new Date()),
                          )}
                        </TableCell>
                        <TableCell align="center">{vaccinationTime}</TableCell>
                        <TableCell align="center">
                          {isConcluded ? (
                            <span className={classes.concluded}>Atendido</span>
                          ) : (
                            <span className={classes.notConcluded}>
                              Aguardando
                            </span>
                          )}
                        </TableCell>
                        <TableCell align="center">
                          {conclusion || '-'}
                        </TableCell>
                        <TableCell align="center">
                          {isConcluded ? (
                            <Button disabled>ConcluÃ­do</Button>
                          ) : (
                            <Button onClick={() => handleDialogOpen(_id)}>
                              Concluir
                            </Button>
                          )}
                        </TableCell>
                      </TableRow>
                    ),
                  )}
                </TableBody>
              </Table>
            </TableContainer>
          ) : (
            <div>
              <Typography>Nenhum agendamento encontrado.</Typography>
              <Image
                imageStyle={{ objectFit: 'contain', marginTop: '50px' }}
                src={undrawNoData}
              />
            </div>
          )}
        </Grid>
      </Grid>
      <Dialog
        fullWidth
        maxWidth="sm"
        open={showDialog}
        onClose={handleDialogClose}
      >
        <form onSubmit={event => handleConclusionSubmit(event)}>
          <DialogTitle id="form-dialog-title">Concluir Atendimento</DialogTitle>
          <DialogContent>
            <TextField
              multiline
              onChange={event => setConclusionInput(event.target.value)}
              autoFocus
              margin="dense"
              id="name"
              label="Detalhes do atendimento"
              type="text"
              fullWidth
            />
          </DialogContent>
          <DialogActions>
            <Button type="submit">Enviar</Button>
          </DialogActions>
        </form>
      </Dialog>
    </Page>
  );
}
