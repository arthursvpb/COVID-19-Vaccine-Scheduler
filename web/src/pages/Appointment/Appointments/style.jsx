import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
  paper: {
    display: 'flex',
    padding: '70px',
    maxWidth: '700px',
  },
  grid: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-end',
    '& > *': {
      display: 'flex',
      justifyContent: 'center',
    },
  },
  title: {
    marginBottom: '20px',
    marginTop: '100px',
  },
  concluded: {
    color: '#43703f',
    borderRadius: '4px',
    textAlign: 'center',
    padding: '5px',
    backgroundColor: '#bfffb9',
  },
  notConcluded: {
    color: '#34646d',
    borderRadius: '4px',
    textAlign: 'center',
    padding: '5px',
    backgroundColor: '#c1f6ff',
  },
});

export default useStyles;
