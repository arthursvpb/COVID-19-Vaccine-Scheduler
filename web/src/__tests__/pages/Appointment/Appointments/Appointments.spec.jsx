/* eslint-disable no-undef */
import React from 'react';

import { render } from '@testing-library/react';
import Appointments from '../../../../pages/Appointment/Appointments';

test('renders correctly', () => {
  const { asFragment } = render(<Appointments />);

  expect(asFragment()).toMatchSnapshot();
});
