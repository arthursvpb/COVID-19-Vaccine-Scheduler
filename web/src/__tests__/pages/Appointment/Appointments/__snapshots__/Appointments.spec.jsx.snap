// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-6 MuiContainer-maxWidthLg"
    style="flex-direction: column; justify-content: flex-start;"
  >
    <div
      class="MuiGrid-root makeStyles-grid-2 MuiGrid-container"
    >
      <div
        class="MuiGrid-root makeStyles-title-3 MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <div
              class="MuiFormControl-root MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                data-shrink="true"
                for="dateFilter"
                id="dateFilter-label"
              >
                Data da vacinação
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                    />
                  </svg>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  id="dateFilter"
                  name="dateFilter"
                  style="text-align: center; font-size: 20px;"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-9 PrivateNotchedOutline-legendNotched-10"
                  >
                    <span>
                      Data da vacinação
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Nenhum agendamento encontrado.
          </p>
          <div
            style="background-color: rgb(255, 255, 255); position: relative;"
          >
            <img
              src="undraw_no_data.svg"
              style="width: 100%; height: 100%; position: absolute; object-fit: contain; top: 0px; left: 0px; opacity: 0; transition: filterBrightness 2250ms cubic-bezier(0.4, 0.0, 0.2, 1),
        filterSaturate 3000ms cubic-bezier(0.4, 0.0, 0.2, 1),
        opacity 1500ms cubic-bezier(0.4, 0.0, 0.2, 1); margin-top: 50px;"
            />
            <div
              style="width: 100%; height: 100%; position: absolute; top: 0px; left: 0px; display: flex; align-items: center; justify-content: center; pointer-events: none;"
            >
              <div
                class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                role="progressbar"
                style="width: 48px; height: 48px;"
              >
                <svg
                  class="MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
